### ISSUE TITLE
A compromise on bangs, you can only add so many engines

### ISSUE THREAD

**Issue by thiswillbeyourgithub** [searx/searx#1146]
Hey devs !

I love searx but I think there is something *fundamentaly* wrong with the way bangs work in searx.

You see, in duckduckgo (and qwant) the idea is basically that pretty much every site can be added as a bang. Hence ddg [claims](https://duckduckgo.com/bang) (today) 9980 different bangs ! This is in my opinion, a huge plus in terms of browsing experience.

But it seems that searx only use as "bangs" the engines that are added and activated in the settings. And having 10 000 engines (and choosing which to activate) is a terrible idea for a metaserach engine.

So can you devs clarify what the end objective is with bangs ? Is it just to display the results from one engine at a time ? or is it to get instant results from another website ? (ddg uses the second)

I think the best compromise would be the following : 

* you keep the system of ":language" returns results from a specific language

* you use "!" to invoke bangs (with redirection to the website) from qwant or duckduckgo (the one the user specified in the settings, in an opt in fashion, kind of like with *autocomplete*)

* you use "&" or any other symbol you see fit to call a specific engine of searx (this is currenty what happens with "!")

I know it would be an hard transition but I think it's the best way forward.

I read issue #870 where it is explained by @kvch that redirection is dangerous (I perfectly valid point) and that's why I think it should be "opt in" (= the user has to activate it himself).

Searx is a great service but I can't wrap my mind arround why I would need to add and activate "wordreference" or "urbandictionnary" or 'imdb" as engines to have instant translation when they would be useless when aggregated with other sites.

I hope I'm clear, don't hesitate to ask me clarification otherwise :)
And don't take it the wrong way, i LOVE searx and I'm a huge advocate for it <3

**Comment by asciimoo** [comment]
This can be a useful addition, but I wouldn't change the api. It could be `!!`.

**Comment by thiswillbeyourgithub** [comment]
@asciimoo just for the record, you mean that `!!` should send to the website of the bang?

for example : `!!imdb test` should lend you straight to the results of imdb on their website?

To put it in another way : `!!imdb test` on searx would be the same as `!imdb test` on ddg/qwant ?

**Comment by asciimoo** [comment]
@thiswillbeyourgithub yes

**Comment by RedSoxFan04** [comment]
asciimoo, I don't know why, but that doensn't work when I try it.

**Comment by kvch** [comment]
@RedSoxFan04 What does not work when you try it? If you mean redirecting to the page by typeing `!!{enginename}`, it's because this feature is not implemented.

**Comment by shumvgolove** [comment]
Why use ```!!``` instead of ```!```? The writing becomes a little longer, but there is no practical benefit. 

Anyway, this feature is a very useful thing I'd like to see in searx.

**Comment by thiswillbeyourgithub** [comment]
@shumvgolove as a way to differentiate `!` from `!!`. It was a mere proposal but it could work. The idea is that sometimes you want to go straight to a website (like the bangs of qwant or duckduckgo) and sometimes you just want the results returned from a website. I explained this more precisely above :).

**Comment by shumvgolove** [comment]
Yes, I'm sorry. Originally did not know about `!` Only after reading the documentation I understood what was the point.

**Comment by dessalines** [comment]
I closed the previous issue and am posting my comment here since its the same idea as @thiswillbeyourgithub 

Instead of bangs searching the metasearch engines, it would be wonderful if bangs were **redirects** to use that sites own search engine. Duckduckgo has ~12k bangs... this is something that wouldn't be feasible or necessary for a metasearchengine.


You could simply added a json file that looks something like:

```
bangs: [
{
  name: "amazon",
  shortcuts : ["a", "amazon"],
  url: "https://www.amazon.com/s/?url=search-alias%3Daps&field-keywords=%s"
  ...
},
```

I'm switching from duckduckgo, and there are many bangs that I can no longer use now: `!idope` or `!tpb` for torrents, `!w` for wikipedia, `!yt` for youtube, `!a` for amazon, `!gh` for github...

The nice thing is that users could just add new urls to this json file in a clean way. 

edit: Also the list of bangs should all be in one place, right now its spread across many, I had no idea `!stackoverflow` worked until I randomly tried it.

**Comment by jfowl** [comment]
I do support `!!bang` for redirections and `!bang` for direct results in searx. It offers the best backwards compatibility.
In order to not get this too mixed up, I would propose to integrate this feature into the settings.yml (or if it ever get's changed into engines.yml):
1) Making the `shortcut` setting an array.
2) Adding a seperate `bang_search_url` setting.

We would need to hook in here: https://github.com/asciimoo/searx/blob/master/searx/webapp.py#L490-L507

**Comment by bat999** [comment]
Hi
If we are going to use searx bangs like this...
!y for youtube, !a for amazon etc
Or !!y for youtube, !!a for amazon etc
I think it's a good idea.

But we need to be careful because there are lots of duplicates among the different search engines.
We might make things even more complicated.
And we'll need a glossary for them in the wiki somewhere.
(Like **dessalines** said above)

For example...

!di with duckduckgo search engine leads to dictionary.com
Very useful

!di with qwant search engine leads to directindustry.fr
Not very useful

**Comment by Madis0** [comment]
Why not just let the user choose the bangs provider?

**Comment by bat999** [comment]
Or maybe we should align with just one search engine's bangs.
Make the decision up front.

Duckduckgo for instance...
Because (I think) it was Duckduckgo that "pioneered" bangs for search engines in 2008.
Now there are "13,505 bangs and counting".

We don't need to maintain a glossary of our own, merely mention bangs in a wiki and link to Duckduckgo's website...
https://duckduckgo.com/bang

There's no shame in that.
:-)

**Comment by return42** [comment]
FWIW to mention: searx is about privacy .. with morty it also includes a result proxy. Users must validate the result and decide if it can be trusted or not. bangs in searx are for selecting one specific engine, not to jump instant to another page like ddg does

Jumping instant to another page is out of the scope of privacy, means it is a complete different application. We can add such a application, but the we can also add other useful stuff like a calculator and a dictionary .. or even a cooking book.

**Comment by return42** [comment]
Sorry, there was a typo in my last post: I mean "FWIW" leaving the leaving the trailing W was not my intention. Again sorry.

**Comment by bat999** [comment]
>  bangs in searx are for selecting one specific engine, not to jump instant to another page like ddg does
> 
OK, I understand now.

**Comment by bat999** [comment]
Hi
I've had some further thoughts about DuckDuckGo !bangs with Searx.
So I will make this last comment then leave it alone. 

Suppose we want to search for a term **foo** with a website such as Wikipedia (or imdb.com or dictionary.com etc...)

We could...

1. Visit the website www.wikipedia.org
Enter the search term **foo**.
2. Visit the website www.duckduckgo.com
Enter the search term **!w foo**.
3. Set DuckDuckGo as the browser's default search engine.
Enter the search term **!w foo** in the browser omnibox.

This third method is quick and very user-friendly. 
But it seems that there are two reasons why the third method can't or won't be used by Searx as the defualt search engine.

1. Searx already uses "**!**" symbol for something else (to specify a particular search engine).
2. The privacy or security of Searx could be affected.

If these two reasons are genuine blockers then this issue might as well be "won't fix". 
It would have been beneficial (imho) if Searx would honour the DuckDuckGo !bangs when set as the browser's default search engine.
 But it is what it is.

**Comment by Madis0** [comment]
From what I understand, the owner of the repo is already supportive of this, just that: 

- it must differ from current bangs (e.g. !! vs !) 
- they are not willing to add it by themselves (but could consider a PR)

**Comment by kvch** [comment]
@Madis0 This issue is on the Project board of the v1.0.0, so it is highly likely that the feature will be implemented before searx 1.0 is released.

**Comment by return42** [comment]
>> We can add such a application, but then we can also add other useful stuff like a calculator and a dictionary .. or even a cooking book.
>> ...
>>  bangs in searx are for selecting one specific engine, not to jump instant to another page like ddg does

> OK, I understand now.

To clarify, I am not totally against such features, but I want to see it in a wider discussion, there are also other nice features we can add to our search syntax, e.g see PR #384 which adds a calculator .. and if one wants to add a cooking book ..why not.

The problem I see is not the addition of nice tools. The problem I want to solve first is: How can we extend our search syntax to be scalable through a wide range of nice tools.

**Comment by jorsn** [comment]
> How can we extend our search syntax to be scalable through a wide range of nice tools.

I know of a nice elisp interpreter which would also provide psochological aid to those developping parenthesophobia. ;-)

Maybe it would really make sense to define the scope of searx, and then define (or adopt) a suitable query language, because it makes no sense to become yet another general language interpreter (call it spython?) with some small core functionality attached.

One could have a general syntax of `!<function><space><query>` where function interprets the query. Maybe allow sub-functions, which filter their input for the first function, e.g. `!f1 query1 !f2 query2`, maybe with a handful of special symbols as operators (`!`, `!!`, `@`, `*`, `"`,…). The top-level function must return a query type (or a result type) interpreted by searx.

If you want it really flexible, introduce grouping and substitute `!` by `\`: `\f1{query1 \f2{query2} \f3{query3}}`. Introduce function application, lambda abstraction, …, types, kinds, universes… where to stop? What's the scope?

**Comment by bat999** [comment]
Hi
Some people might definitely not want to use bangs with Searx.
For reasons of privacy or security etc.
That's understandable.

Maybe we could make them optional with a setting in Searx's "Preferences".
  
```
1) Enable !!bang syntax.
Or
2) Disable !!bang syntax.
```

If bangs are disabled then Searx would treat them as literal searches, as if they were in quotes.
`"!!w foo"`

Eventually it might be extended to a 3-way coice...

```
1) Enable !!bang syntax with Searx's database.
Or
2) Enable !!bang syntax with DuckDuckGo's database.
Or
3) Disable !!bang syntax.
```

If a user chose Searx's database he would have a safe subset.
If a user chose DuckDuckGo's database he would have a whole set but without much (or any) quality assurance.

**Comment by asciimoo** [comment]
>  Some people might definitely not want to use bangs with Searx.

If somebody doesn't want to use external bangs, then simply do not enter a query using external bangs. What would be the advantage of disabling it completely?

**Comment by bat999** [comment]
> 
> If somebody doesn't want to use external bangs, then simply do not enter a query using external bangs.

I'm OK with that.
:-)

**Comment by dalf** [comment]
I agree it is disturbing because ddg's users are used to the ```!``` syntax, but I would enforce the current syntax:
* !  searx bang (Searx's database.)
* !! external bang (with DuckDuckGo's database)

Even for a private instance I think it is a problem to let the people chooses:
* someone uses a private searx instance and get use to type (and don't think about it):
  * ```!!ddg <something>``` to get the instant answer from the ddg.
  * ```!yandex <something>``` to get result from yandex.
* for some reason the private instance shutdowns.
* this user pick a public instance, everything look the same, but the user types ```!yandex ...```, the instance is configured the other way around: the privacy is broken (the purpose of searx).

So in the end, the user has to check which one is which before using it, and it is counter intuitive because when you get use to a syntax it is difficult to change.

Perhaps out of topic: it like a touch screen in a car, you have to look at screen to understand what it does, but you don't look anymore at the road. Compare to physical button: you don't have to look at it, because your hand knows where it is.

It's only my opinion.

**Comment by bat999** [comment]
> 
> * !  searx bang (Searx's database.)
> * !! external bang (with DuckDuckGo's database)
> 
That's OK
One exclamation point to denote an engine.
Two exclamation points to denote a bang.

Probably Searx users (including me) will be able to remember to use double exclamation points for bangs.

But what is being worked on seems to be a !!bang database that is not the full database used by DuckDuckGo.
It seems to be a subset of "clean", "curated", "safe" bangs instead of the the whole "wild west" set from DuckDuckGo.
At DuckDuckGo there are "Currently **13,505** bangs and counting".

But the "safe" Searx !!bang subset adds complications...

* Do we know that a bang that we previously used with DuckDuckGo is included in the "safe" subset?

* Do we know that a bang that we previously used with DuckDuckGo leads to the same place with the "safe" subset?

I gave an example in a previous post that... 
`!di foo` leads to **dictionary.com** with **DuckDuckGo**.
`!di foo` leads to **directindustry.fr** with **Qwant**.

**Comment by dalf** [comment]
Technically, it is possible to get exactly the same bangs ```duckduckgo.com/bang.v255.js``` (the format is different from https://raw.githubusercontent.com/asciimoo/searx/master/searx/data/bangs.json ). 

A similar autocomplete to ddg would be helpful (with categories and subcategories).

**Comment by asciimoo** [comment]
I'm closing this issue, since the original issue is implemented by #2027 . Feel free to reopen if there are unresolved topics.
